cscs-checks/microbenchmarks/gpu/gpu_burn/src/gpu_burn.cu

```diff
@@ -39,553 +42,369 @@
-                // printf("proc'd: ");
-                // for (size_t i = 0; i < clientCalcs.size(); ++i) {
-                // printf("%d (%.0f Gflop/s) ", clientCalcs.at(i), clientGflops.at(i));
-                // if (i != clientCalcs.size() - 1)
-                // printf("- ");
-                // }
-                // printf("  errors: ");
-                for (size_t i = 0; i < clientErrors.size(); ++i) {
-                    std::string note = "%d ";
-                    // if (clientCalcs.at(i) == -1)
-                    // note += " (DIED!)";
-                    // else if (clientErrors.at(i))
-                    // note += " (WARNING!)";
-
-                    // printf(note.c_str(), clientErrors.at(i));
-                    // if (i != clientCalcs.size() - 1)
-                    // printf("- ");
-                }
-                // printf("  temps: ");
-                // for (size_t i = 0; i < clientTemp.size(); ++i) {
-                // printf(clientTemp.at(i) != 0 ? "%d C " : "-- ", clientTemp.at(i));
-                // if (i != clientCalcs.size() - 1)
-                // printf("- ");
-                // }
-
-                fflush(stdout);
-
-                if (nextReport < elapsed) {
-                    nextReport = elapsed + 2.0f;
-                    // printf("\n\tSummary at:   ");
-                    // fflush(stdout);
-                    // system("date"); // Printing a date
-                    // fflush(stdout);
-                    // printf("\n");
-                    // printf("\t(checkpoint)\n");
-                    for (size_t i = 0; i < clientErrors.size(); ++i) {
-                        if (clientErrors.at(i))
-                            clientFaulty.at(i) = true;
-                        clientErrors.at(i) = 0;
-                    }
-                }
-            }
-
-            // Checking whether all clients are dead
-            bool oneAlive = false;
-            for (size_t i = 0; i < clientCalcs.size(); ++i)
-                if (clientCalcs.at(i) != -1)
-                    oneAlive = true;
-            if (!oneAlive) {
-                fprintf(stderr, "\n\nNo clients are alive!  Aborting\n");
-                exit(123);
-            }
-
-            if (startTime + runTime < thisTime)
-                break;
-    }
+    // Here burn is a switch that holds and breaks the work done by the slave threads.
+    burn = false;
 
-    // printf("\nKilling processes.. ");
-    fflush(stdout);
-    for (size_t i = 0; i < clientPid.size(); ++i)
-        kill(clientPid.at(i), 15);
+    XGetDeviceCount(&devCount);
+    std::vector<std::thread> threads;
 
-    // kill(tempPid, 15);
-    // close(tempHandle);
+    // Print device count
+    printf("[%s] Found %d device(s).\n", hostname, devCount);
 
-    while (wait(NULL) != -1);
-    printf("Node %s:\n", hostname);
+    // All the burn info is stored in instances of the BurnTracker class.
+    BurnTracker * trackThreads = new BurnTracker[devCount];
```

Comment:
I think you are leaking this, because you are constructing again `BurnTracker`s in the loop below with the placement new. Either make this an array of pointers to `BurnTracker`s or dummy construct them here and initialize them with an explicit method that you would call below.

Reply:
Fixed. I've gone for the array of pointers approach.